### ISSUE TITLE
Azure PostgreSQL <username@hostname> Not Supported

### ISSUE THREAD

**Issue by danielcbright** [chef/chef-server#1559]
## Expected Behavior
I expect to be able to configure my `chef-server.rb` file to use an external PostgreSQL database as outlined in Chef documentation like so:

```
postgresql['db_superuser'] = 'chefadmin@azurechefhost' # Postgres user you created
postgresql['db_superuser_password'] = 'somepw' # That user's password
postgresql['external'] = true # Must be set to true
postgresql['port'] = 5432 # Only required if you change the default port
postgresql['vip'] = 'azurechefhost.postgres.database.azure.com' # DNS entry from Azure
```
And then be able to run `chef-server-ctl` reconfigure on a new Chef Server installation to configure my new Chef Server instance to talk to my Azure PostgreSQL backend. This should result in a successful Chef Server configuration/deployment.

## Current Behavior
The current results are not as expected however. Due to the fact that Azure PostgreSQL requires all usernames to be in the format `<username@hostname>`, apparently Ruby code that reads the `postgresql['db_superuser']` entry doesn't interpret the `@` symbol properly and thus doesn't authenticate and perform the install, here is the full error stack:

```
[root@centoschef opscode]# chef-server-ctl reconfigure
Starting Chef Client, version 12.21.31
resolving cookbooks for run list: ["private-chef::default"]
Synchronizing Cookbooks:
  - private-chef (0.1.1)
  - openssl (8.1.2)
  - enterprise (0.11.0)
  - runit (4.0.4)
  - yum-epel (2.1.2)
  - compat_resource (12.19.1)
  - packagecloud (0.3.0)
Installing Cookbook Gems:
Compiling Cookbooks...
/var/opt/opscode/local-mode-cache/cookbooks/packagecloud/resources/repo.rb:10: warning: constant ::Fixnum is deprecated
Recipe: private-chef::default
  * directory[/etc/opscode] action create (up to date)
  * directory[/etc/opscode/logrotate.d] action create (up to date)
[2018-08-23T14:39:36+00:00] FATAL:

-----------------------------------------------------------------------
CSPG999: FATAL:  Invalid Username specified. Please check the Username and retry connection. The Username should be in <username@hostname> format.
FATAL:  Invalid Username specified. Please check the Username and retry connection. The Username should be in <username@hostname> format.
-----------------------------------------------------------------------
```

## Steps to Reproduce (for bugs)

1. Start with a fresh Azure PostgreSQL DB and fresh Centos 7.4 VM.
2. Edit your `chef-server.rb` file as shown in the example above to match your AzurePostgreSQL settings.
3. Run `chef-server-ctl reconfigure` to setup your Chef Server, and you will get the shown results.

## Your Environment
* Chef Server Version: 12.21.31
* Total/free RAM and disk space: 8GB/20GB (Test Server)
* Operating System and Version: Centos 7.4 w/Updates
* If upgrading, previous Chef Server version: N/A
* Running in a container? No (Azure)

**Comment by danielcbright** [comment]
I also want to note that I have used [PGBouncer](https://pgbouncer.github.io/) to attempt and proxy the username information and I was successfully able to run the install. However I'm not sure this is a good long-term solution as the way it works is it maps all users from the Chef-side to a singe user in PostgreSQL. If that's seen as an acceptable workaround then I'll post my config, otherwise I don't want to muddy the waters with a workaround that isn't extensively tested.

**Comment by nhudacin** [comment]
We just hit this ourselves. Commenting here to follow the issue.

**Comment by markan** [comment]
Most likely we're not handling the error case correctly in the connectivity check here: https://github.com/chef/chef-server/blob/5c7c988cba529c4794b21faa9c6bd135f37f47c8/omnibus/files/private-chef-cookbooks/private-chef/libraries/preflight_postgres_validator.rb#L110 
We test for an invalid user with the symbol :invalid_user, and I suspect azure pgsql is flagging the format of the username, and  we're not set up to handle the error properly.  Adding an extra clause to handle this would be easy enough I think. Probably need to do some unit tests to cover for the fact that we don't currently don't test this particular configuration in the pipeline.

**Comment by zanecodes** [comment]
We've been working with @markan on this and there's a lot of non-obvious stuff going on with this.

To sum it up, Azure Database for PostgreSQL has a gateway in front of the actual database servers that forwards connections from `username@hostname` to `hostname` as `username`.
This means that once the connection is established, you will actually be connected as `username`, not `username@hostname`, and any database queries involving users should just use `username` (e.g. granting permissions).

The `private-chef` cookbook assumes the same username for both purposes (connections and queries), but once it configures the services, they only use the username for connections, they don't seem to perform any queries with them.

In order to support Azure PostgreSQL, the cookbook will have to be modified to allow overriding the connection username separately from the database username, but defaulting them to be the same for backwards compatibility. This will probably require a new attribute for each database user to be added (something like `node['private-chef']['opscode-erchef']['db_sql_user']`), possibly renaming the existing attributes to clarify their usage, and replacing all uses of the existing attributes in database queries with the new attributes. Not sure how to fully test the scenario this supports though.

This isn't quite enough, as Postgres queries involving an `@` in the username will fail with invalid syntax errors unless the username is surrounded by double quotes, and not all such queries in the `private-chef` cookbook do so. Adding tests for this should be relatively straightforward.

Finally, Azure Postgres enforces SSL by default. This can be disabled, but that's less than ideal. `epgsql`, the database library used by the Chef Erlang services, [does seem to support SSL connections](https://github.com/epgsql/epgsql#connect), and `sqerl`, the abstraction layer that wraps it, [added support for setting additional connection options](https://github.com/chef/sqerl/blob/master/src/sqerl_pgsql_client.erl#L192) [via the `sys.config`](https://github.com/chef/sqerl/blob/master/src/sqerl_config_env.erl#L39). We should probably add an attribute that allows for overriding the ssl mode for each service (or maybe just one for all of them) and pass that down to the `sqerl.db_options.ssl` key in each service's `sys.config`.

So the changes needed in `private-chef` to support Azure Database for PostgreSQL will be at least:
- Correct username quoting in queries
- Separate connection usernames from db usernames
- Add attributes to override only connection usernames

And ideally also:
- Add support for overriding the sslmode with an attribute

**Comment by PrajaktaPurohit** [comment]
@danielcbright This issue is currently being worked on and is close to getting merged. It should be available in the next release of Chef-Infra-Server. Thank you for your patience on this!

**Comment by danielcbright** [comment]
@PrajaktaPurohit just to confirm, are we currently waiting for test cases to be built out for this so it can be tested and merged?

**Comment by PrajaktaPurohit** [comment]
@danielcbright yes that is correct. #1728 has the changes. We are currently working on testing it with the different scenarios. (Description is updated with the scenarios currently tested with) We can merge it as soon as all of the scenarios pass. We have a release coming up the month and another one in December.

**Comment by danielcbright** [comment]
Looks like this issue might be closed, do we have a release that this has been added to yet? @PrajaktaPurohit ?

**Comment by PrajaktaPurohit** [comment]
Indeed, this change has shipped as a part of 13.2.0 release of the Chef Infra Server. I will include the version shipped before closing henceforth. Sorry for the trouble!
